benchmark/utils.cpp

```diff
@@ -0,0 +1,231 @@
+                config->obj_func = std::string(optarg);
+                break;
+            case 'f':  // output file name
+                config->out_file = std::string(optarg);
+                break;
+            case 's':  // solution output name
+                config->solution_file = std::string(optarg);
+                break;
+            case 'd':  //  dimension
+                int dim;
+                if(sscanf(optarg, "%i", &dim) != 1) {
+                    fprintf(stderr, "invalid arg '%s': must be an integer\n", opt);
+                    exit(EXIT_FAILURE);
+                }
+                config->dimension = dim;
+                break;
+            case 'p':  // population
+                int population;
+                if(sscanf(optarg, "%i", &population) != 1) {
+                    fprintf(stderr, "invalid arg '%s': must be an integer\n", optarg);
+                    exit(EXIT_FAILURE);
+                }
+                config->population = population;
+                break;
+            case 'n':  // n_iterations
+                int n_iter;
+                if(sscanf(optarg, "%i", &n_iter) != 1) {
+                    fprintf(stderr, "invalid arg '%s': must be an integer\n", optarg);
+                    exit(EXIT_FAILURE);
+                }
+                config->n_iterations = n_iter;
+                break;
+            case 'y':  // min_position
+                int min_pos;
+                if(sscanf(optarg, "%i", &min_pos) != 1) {
+                    fprintf(stderr, "invalid arg '%s': must be an integer\n", optarg);
+                    exit(EXIT_FAILURE);
+                }
+                config->min_position = min_pos;
+                break;
+            case 'z':  // max_position
+                int max_pos;
+                if(sscanf(optarg, "%i", &max_pos) != 1) {
+                    fprintf(stderr, "invalid arg '%s': must be an integer\n", optarg);
+                    exit(EXIT_FAILURE);
+                }
+                config->max_position = max_pos;
+                break;
+
+            case 'h':
+            default:
+                fprintf(stderr, USAGE);
+                exit(EXIT_FAILURE);
+        }
+    }
+}
+
+/**
+   Generate a random number between 0 and 1.
+*/
+double randf() {
+    return (double)rand() / (double)RAND_MAX;
+}
+
+/**
+   Build an array of size `ARRAY_SIZE` and fill it with random numbers.
+*/
+double* build() {
+    double* x = (double*)malloc(ARRAY_SIZE * sizeof(double));
+    for(size_t i = 0; i < ARRAY_SIZE; ++i) {
+        x[i] = randf();
+    }
+    return x;
+}
+
+
+void print_file(std::string file_path) {
+
+    std::string text;
+    std::ifstream ifs(file_path);
+
+    while(!ifs.eof())
+    {
+        std::getline(ifs, text);
+        std::cout << "" << text << "\n" ;
+    }
+}
+
+
+void print_git_version() {
+    std::cout << "# Git"  << std::endl;
+    std::cout << "git version - " << GIT_VERSION << std::endl;
```

Comment:
How do you obtain those values? The `GIT_VERSION` etc?

Reply:
They are passed in via the Makefile. See benchmark Makefile lines 12-14. __However__, now that we do everything in python, we might as well port this functionality over to the python code since I think this is kinda ugly here and will be much nicer over in python. Since C++ and Python are in the same git repo, the hashes should still be valid. Opinions?

Reply:
Mhh yeah ... technically I like the fact that it is in the makefile, since the compilation time is the important one for the versioning. However, it might be more flexible in python that's true. I guess we will never encounter a problem with having an old executable around if we compile in the python script before launching the experiments... I am down for either, do what you prefer